yamux/src/frame.rs

```diff
@@ -27,17 +27,59 @@ use self::header::{Flags, Tag, HEADER_SIZE};
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct Frame<T> {
     buffer: Vec<u8>,
-    _marker: std::marker::PhantomData<T>,
+    _marker: PhantomData<T>,
 }
 
 impl<T> Frame<T> {
-    pub(crate) fn new(buffer: Vec<u8>) -> Self {
+    pub(crate) fn new(header: Header<T>) -> Self {
+        let total_buffer_size = HEADER_SIZE + header.len().val() as usize;
```

Comment:
You can't allocate space for body here, it depends on the kind of header, for instance Header<WindowUpdate> encodes credit into its len field without providing a body.